src/Compilers/CSharp/Portable/Binder/ForEachEnumeratorInfo.cs

```diff
@@ -23,10 +21,17 @@ internal sealed class ForEachEnumeratorInfo
         public readonly MethodSymbol CurrentPropertyGetter;
         public readonly MethodSymbol MoveNextMethod;
 
+        // In async-foreach, those two methods will be populated instead of CurrentPropertyGetter and MoveNextMethod
```

Comment:
@OmarTawfik, I took a look at your suggestion to split the `ForEachEnumeratorInfo` class (and possibly its builder) into a hierarchy. But it didn't end up improving the binding or the lowering code, since the most of the code is in common between async and non-async case :-(
I'll leave as-is if that's ok, and I'm happy to add clarifying comments wherever useful.